src/mds/auth/authnode.cpp

```diff
@@ -0,0 +1,262 @@
+    const std::string &in, std::string *out) {
+    std::string authKey;
+    {
+        ReadLockGuard keyMapLG(keyMapMutex_);
+        auto it = keyMap_.find(curve::common::AUTH_ROLE);
+        if (it == keyMap_.end()) {
+            LOG(ERROR) << "DecryptByAuthKey failed, the auth key not found.";
+            return AuthStatusCode::AUTH_KEY_NOT_EXIST;
+        }
+        authKey = it->second.key();
+    }
+    auto ret = Encryptor::AESDecrypt(authKey, ZEROIV, in, out);
+    if (ret != 0) {
+        LOG(ERROR) << "DecryptByAuthKey failed, in = " << in;
+        return AuthStatusCode::AUTH_DECRYPT_FAILED;
+    }
+    return AuthStatusCode::AUTH_OK;
+}
+
+AuthStatusCode AuthNodeImpl::EncryptByAuthKey(
+    const std::string &in, std::string *out) {
+    std::string authKey;
+    {
+        ReadLockGuard keyMapLG(keyMapMutex_);
+        auto it = keyMap_.find(curve::common::AUTH_ROLE);
+        if (it == keyMap_.end()) {
+            LOG(ERROR) << "EncryptByAuthKey failed, the auth key not found.";
+            return AuthStatusCode::AUTH_KEY_NOT_EXIST;
+        }
+        authKey = it->second.key();
+    }
+    auto ret = Encryptor::AESEncrypt(authKey, ZEROIV, in, out);
+    if (ret != 0) {
+        LOG(ERROR) << "EncryptByAuthKey failed, in = " << in;
+        return AuthStatusCode::AUTH_DECRYPT_FAILED;
+    }
+    return AuthStatusCode::AUTH_OK;
+}
+
+AuthStatusCode AuthNodeImpl::AddKey(const std::string &encKey) {
+    std::string decKeyStr;
+    Key key;
+    auto ret = DecryptByAuthKey(encKey, &decKeyStr);
+    if (AuthStatusCode::AUTH_OK != ret) {
+        return ret;
+    }
+
+    if (!key.ParseFromString(decKeyStr)) {
+        LOG(ERROR) << "Add key failed, parse key failed, decrypt key = "
+                   << decKeyStr;
+        return AuthStatusCode::AUTH_DECRYPT_FAILED;
+    }
+
+    WriteLockGuard keyMapLG(keyMapMutex_);
+    auto it = keyMap_.find(key.id());
+    if (it != keyMap_.end()) {
+        LOG(WARNING) << "Add key but key exist and will update it, oldkey = "
+                     << keyMap_[key.id()].DebugString() << ", newkey = "
+                     << key.DebugString();
+    } else {
+        LOG(INFO) << "Add key, key = " << key.DebugString();
+    }
+    keyMap_[key.id()] = key;
+    if (!storage_->StorageKey(key)) {
+        return AuthStatusCode::AUTH_STORE_KEY_FAILED;
+    }
+    return AuthStatusCode::AUTH_OK;
+}
+
+AuthStatusCode AuthNodeImpl::DeleteKey(const std::string &encKeyId) {
+    std::string decKeyId;
+    auto ret = DecryptByAuthKey(encKeyId, &decKeyId);
+    if (AuthStatusCode::AUTH_OK != ret) {
+        return ret;
+    }
+
+    WriteLockGuard keyMapLG(keyMapMutex_);
+    LOG(INFO) << "Delete key, key id = " << decKeyId;
+    keyMap_.erase(decKeyId);
+    if (!storage_->DeleteKey(decKeyId)) {
+        return AuthStatusCode::AUTH_DELETE_KEY_FAILED;
+    }
+    return AuthStatusCode::AUTH_OK;
+}
+
+AuthStatusCode AuthNodeImpl::UpdateKey(const std::string &encKey) {
```

Comment:
it seems `AddKey` has same logic with `UpdateKey`

Reply:
The logic is similar but not identical, if it does not exist when updatekey will be added

Reply:
so that, can you just add another parameter to indicate whether to update or add instead of two similar functions?